### DISCUSSION TITLE
How to use copy.deepcopy() in place of clone()?

### DISCUSSION THREAD

**Discussion by miserva** [pyuvm/pyuvm#105] [Q&A] [answered]
According to #84 it's recommended to use copy module instead of `clone()`, and it would definitely save some effort. However, when I tried to replace `clone()` with `copy.deepcopy()` in the common scenario where a driver copies a request item to create a response, I run into an issue.

```python
class my_driver(uvm_driver):
    ...
    async def drive(self):
        while True:
            req = await self.seq_item_port.get_next_item()
            resp = copy.deepcopy(req)
   ...
```
With the code above I get this error:
```
# Traceback (most recent call last):
#   File ".../src/my_driver.py", line 23, in run_phase
#     await self.drive()
#   File ".../src/my_driver.py", line 40, in drive
#     resp = copy.deepcopy(req)
#   File ".../lib/python3.7/copy.py", line 180, in deepcopy
#     y = _reconstruct(x, memo, *rv)
#   File ".../lib/python3.7/copy.py", line 280, in _reconstruct
#     state = deepcopy(state, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 240, in _deepcopy_dict
#     y[deepcopy(key, memo)] = deepcopy(value, memo)
#   File ".../lib/python3.7/copy.py", line 180, in deepcopy
#     y = _reconstruct(x, memo, *rv)
#   File ".../lib/python3.7/copy.py", line 280, in _reconstruct
#     state = deepcopy(state, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 240, in _deepcopy_dict
#     y[deepcopy(key, memo)] = deepcopy(value, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 215, in _deepcopy_list
#     append(deepcopy(a, memo))
#   File ".../lib/python3.7/copy.py", line 180, in deepcopy
#     y = _reconstruct(x, memo, *rv)
#   File ".../lib/python3.7/copy.py", line 280, in _reconstruct
#     state = deepcopy(state, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 220, in _deepcopy_tuple
#     y = [deepcopy(a, memo) for a in x]
#   File ".../lib/python3.7/copy.py", line 220, in <listcomp>
#     y = [deepcopy(a, memo) for a in x]
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 240, in _deepcopy_dict
#     y[deepcopy(key, memo)] = deepcopy(value, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 247, in _deepcopy_method
#     return type(x)(x.__func__, deepcopy(x.__self__, memo))
#   File ".../lib/python3.7/copy.py", line 180, in deepcopy
#     y = _reconstruct(x, memo, *rv)
#   File ".../lib/python3.7/copy.py", line 280, in _reconstruct
#     state = deepcopy(state, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 240, in _deepcopy_dict
#     y[deepcopy(key, memo)] = deepcopy(value, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 240, in _deepcopy_dict
#     y[deepcopy(key, memo)] = deepcopy(value, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 215, in _deepcopy_list
#     append(deepcopy(a, memo))
#   File ".../lib/python3.7/copy.py", line 180, in deepcopy
#     y = _reconstruct(x, memo, *rv)
#   File ".../lib/python3.7/copy.py", line 280, in _reconstruct
#     state = deepcopy(state, memo)
#   File ".../lib/python3.7/copy.py", line 150, in deepcopy
#     y = copier(x, memo)
#   File ".../lib/python3.7/copy.py", line 240, in _deepcopy_dict
#     y[deepcopy(key, memo)] = deepcopy(value, memo)
#   File ".../lib/python3.7/copy.py", line 169, in deepcopy
#     rv = reductor(4)
# TypeError: can't pickle coroutine objects
```

If I do:
```python
self.logger.debug(f"type:{type(req)} iscoro:{inspect.iscoroutine(req)}")
```
I get:
```
type:<class 'src.my_seq_item.my_seq_item'> iscoro:False
```
So I don't know why it's saying I'm trying to copy a coroutine object.
I also tried the following and it worked:
```python
class my_driver(uvm_driver):
    ...
    async def foo(self):
        return my_seq_item.create("foo")
    async def drive(self):
        while True:
            foo = await self.foo()
            bar = copy.deepcopy(foo)
    ...
```
So the issue shouldn't be about my sequence item or coroutine return objects in general.
Why do I run into an error here and how would I need to modify my code to make deepcopying the request work?

**Comment by raysalemi** [comment]
Interesting.  I've never seen this.

I would need to see your definition of the `req` class to understand where it's finding a coroutine.  It may be that my recommendation is flawed.

That said, you can get around the problem by overriding `do_copy()` in your sequence item and explicitly copying the fields you want copied.  You would need to create new instances of your objects to get a true deep copy.  

Once you've defined `do_copy()` you can use 

`resp = req.clone()`


# copy module vs. UVM copy

I've gone back and forth on this topic.  My instinct was to leverage the well-tested **copy** module that comes with Python.  But now I'm wondering if it's better advice to have people explicitly override `do_copy()`.

**Reply by miserva** [reply]
I have been using `clone()` so far, but I was looking for ways to improve my code and that's why I tried the copy module.

I got the same error using a dummy sequence item and sequence.
```python
class dummy_seq_item(uvm_sequence_item):
    ...
```
```python
class dummy_sequence(uvm_sequence):
    async def body(self):
        tr = dummy_seq_item.create("dummy_seq_item_tr")
        await self.start_item(tr)
        await self.finish_item(tr)
        resp = await self.get_response()
```
Feeding this sequence to my_driver returned the same error.
I added some debugging code to see if the item passed via sequencer and the item created by `foo()` have any difference:
```
bar var types:{'_obj_name': <class 'str'>, '_initiator': <class 'NoneType'>, 'transaction_id': <class 'int'>, 'start_condition': <class 'cocotb.triggers.Event'>, 'finish_condition': <class 'cocotb.triggers.Event'>, 'item_ready': <class 'cocotb.triggers.Event'>, 'parent_sequence_id': <class 'NoneType'>, 'response_id': <class 'NoneType'>}
req var types:{'_obj_name': <class 'str'>, '_initiator': <class 'NoneType'>, 'transaction_id': <class 'int'>, 'start_condition': <class 'cocotb.triggers.Event'>, 'finish_condition': <class 'cocotb.triggers.Event'>, 'item_ready': <class 'cocotb.triggers.Event'>, 'parent_sequence_id': <class 'int'>, 'response_id': <class 'NoneType'>}
```
On the surface it doesn't show anything, but going layers deeper in introspection I found that `req.finish_condition._pending` has a pending trigger in the list:
```python
finish_pending = []
for item in req.finish_condition._pending:
    finish_pending.append((item, type(item)))
self.logger.debug(f"finish_pending item types:{finish_pending}")
```
```python
finish_pending item types:[(<<Event at 0x2aaac4179f90>.wait() at 0x2aaac41985a0>, <class 'cocotb.triggers._Event'>)]
```
This is the culprit. If I clear the `_pending` list, deepcopying `req` doesn't give an error.

**Reply by raysalemi** [reply]
Interesting. Is the fact that req was not cleared a bug?  Was the sequence item in mid-flight?

**Reply by miserva** [reply]
Ok, I understand what's happening now. This is `cocotb.triggers.Event`:
```python
class Event:
    """Event to permit synchronization between two coroutines.
    Awaiting :meth:`wait()` from one coroutine will block the coroutine until
    :meth:`set()` is called somewhere else.
    """
```
`uvm_sequence_item` has three of these, `start_condition`, `finish_condition` and `item_ready`. They are used to force a certain order of execution between a sequence and a driver.
1. Sequence `start_item()` waits `start_condition`.
2. Driver `get_next_item()` sets `start_condition` and waits `item_ready`.
3. Sequence `finish_item()` sets `item_ready` and waits `finish_condition`.
4. Driver `item_done()` sets `finish_condition`.

Everything is working as intended. It just means that between `get_next_item()` and `item_done()` in the driver, the request item is holding a coroutine object waiting for `item_done()` and thus cannot be deepcopied. Only `clone()` will work here.

**Comment by raysalemi** [comment, marked as answer]
I've got a quick question about your usage if you could email me at ray.salemi@siemens.com.  I'm trying to understand who out there is using **pyuvm** big companies, little companies, individuals?

**Comment by raysalemi** [comment]
Thanks for the investigation!

One thing to remember is that clone() is part of the IEEE 1800.2 standard.
The clone/do_copy() relationship must remain the same.  So users must be
able to override do_copy() to provide their own copy behavior.

That said, users could also use super().do_copy() to preserve the behavior
of the base class.
